### PR TITLE
fix(install): compare settings dev ids only when both stats report one

### LINKED ISSUE #3041: install: Claude settings guards compare fstat dev against lstat dev, blocking every settings write on Windows
## Summary

On Windows, every managed-install path that writes `~/.claude/settings.json` fails closed and can never succeed. Two TOCTOU guards introduced in 2.2.x compare a device id obtained from `fs.fstatSync()` against one obtained from `fs.lstatSync()`. Node on Windows populates `dev` on the descriptor stat but reports `0` on the path stat, so the equality check is unsatisfiable.

This makes `/auto-update` and any `install-apply.js` run that materializes hooks unusable on Windows. It also leaves a stale lock behind on each failure.

This is the same defect class the 2.2.0 changelog records fixing in the `ecc memory` vault guard ("libuv resolved path-based `stat()`/`lstat()` through `GetFileInformationByName` without setting the volume serial, while `fstat()` reported it"). The settings guards reintroduce it. Note the memory-vault fix was framed as a libuv patch-level issue; this reproduces on libuv well past 1.51.0, because the zero here comes from Node's own Windows `lstat` path rather than the runtime's patch level.

## Affected code

**1. `scripts/lib/install/claude-settings.js`, `readSettingsSnapshot()`**

```js
const descriptorStat = fs.fstatSync(descriptor);
...
pathStat = fs.lstatSync(settingsPath);
...
if (
  !descriptorStat.isFile()
  || !pathStat.isFile()
  || pathStat.isSymbolicLink()
  || descriptorStat.dev !== pathStat.dev   // <-- unsatisfiable on Windows
  || descriptorStat.ino !== pathStat.ino
) {
  const error = new Error(`Refusing to read changed Claude settings at ${settingsPath}`);
```

**2. `scripts/lib/install/claude-settings-lock.js`, `sameFileIdentity()`**

```js
function sameFileIdentity(left, right) {
  return left.dev === right.dev && left.ino === right.ino;
}
```

Called with `ownedStats` from `fs.fstatSync(descriptor, { bigint: true })` and `quarantinedStats` from `fs.lstatSync(quarantinePath, { bigint: true })`, so it fails the same way. It is also used by `inspectSettingsLock()`, which compares `fstat` against `lstat` directly.

## Reproduction

No ECC needed:

```js
const fs = require('fs');
const os = require('os');
const path = require('path');

const p = path.join(os.tmpdir(), `ecc-stat-repro-${process.pid}.json`);
fs.writeFileSync(p, '{}');
const fd = fs.openSync(p, fs.constants.O_RDONLY | (fs.constants.O_NOFOLLOW || 0));
const d = fs.fstatSync(fd);
const l = fs.lstatSync(p);
console.log('fstat dev', d.dev, '| lstat dev', l.dev);
console.log('inodes equal', d.ino === l.ino, '| devices equal', d.dev === l.dev);
fs.closeSync(fd);
fs.rmSync(p, { force: true });
```

Output on Windows 11, Node v22.14.0:

```
platform      : win32 | node v22.14.0
O_NOFOLLOW    : undefined
fstat  dev    : 2162558900 | ino 19421773395341796
lstat  dev    : 0 | ino 19421773395341796
inodes equal  : true
devices equal : false   <-- guard compares this
guard rejects : true
```

The inode matches, the file is regular, and it is not a symlink. Only the device id differs, and only because Windows does not populate it on the path stat.

## Observed failure sequence

Running the recorded reinstall on Windows:

1. `install-apply.js ... --enable-hooks` copies files, creates `settings.json.ecc.lock`, then aborts with `Refusing to read changed Claude settings at C:\Users\<user>\.claude\settings.json`.
2. The lock is never released, so the next run reports `Another ECC process is updating Claude settings`, naming a PID that is already dead.
3. After clearing the lock and fixing guard 1, the run writes `settings.json` correctly, then aborts at release with `Refusing to release a changed Claude settings lock`.
4. Each failed attempt leaves another stale lock.

## Suggested fix

Compare devices only when both sides report one. The inode, regular-file, and symlink checks continue to fail closed, so the guard keeps its security property.

```js
// Windows: fs.lstatSync reports dev as 0 while fs.fstatSync reports the real
// device id, so a strict equality check can never pass there. Compare devices
// only when both sides report one.
function sameDeviceId(descriptorStat, pathStat) {
  if (process.platform === 'win32' && (descriptorStat.dev === 0 || pathStat.dev === 0)) {
    return true;
  }
  return descriptorStat.dev === pathStat.dev;
}
```

The lock module needs the BigInt-aware variant, since it stats with `{ bigint: true }`.

Two secondary points worth considering:

- `fs.constants.O_NOFOLLOW` is `undefined` on Windows, so `flags | (fs.constants.O_NOFOLLOW || 0)` silently contributes nothing. The symlink defense there rests entirely on the `isSymbolicLink()` check.
- A guard that aborts after acquiring the lock should release it on the failure path, otherwise every rejection compounds into a stale-lock error that masks the real cause.

## Environment

| | |
|---|---|
| ECC | 2.2.1 (`5064474d`) |
| OS | Windows 11 Home 10.0.26200 |
| Node | v22.14.0 |
| Target | `claude` home install, `hookConsent: enabled` |

## Testing note

`tests/install/` has Windows race fixtures added in `a220947f` ("test(install): exercise settings races safely on Windows"), but they do not catch this, presumably because they construct stats rather than reading them from a real Windows descriptor. A regression test should assert `readSettingsSnapshot()` succeeds against an untampered real file on `win32`.

### PR DESCRIPTION
**Pull request by apetcu**
## What Changed
- `scripts/lib/install/claude-settings.js`: `readSettingsSnapshot()` compares the descriptor `dev` against the path `dev` through a new `sameDeviceId()` helper that treats a missing (`0`/`0n`) device id on either side as "no evidence" instead of "changed file".
- `scripts/lib/install/claude-settings-lock.js`: `sameFileIdentity()` applies the same rule, so lock release and stale-lock inspection stop rejecting the lock the process created itself.
- `tests/lib/claude-settings.test.js`: two new cases. One stubs `fs.lstatSync` to report `dev = 0` (Number and BigInt) and asserts an atomic update succeeds and releases its lock; it fails on `main` with `Refusing to read changed Claude settings` and, after the first guard alone is fixed, with `Refusing to release a changed Claude settings lock`, matching the sequence in #3041. The other stubs a *different* non-zero `dev` and asserts the update is still rejected with `ECC_SETTINGS_CHANGED`.
- `CHANGELOG.md`: Unreleased / Fixed entry, following the format of #2637.

Fixes #3041

## Why This Change
On Windows, Node 22.12-22.16 and 24.0-24.1 (libuv 1.49.0-1.50.x) resolve path-based `lstat()` through `GetFileInformationByName` and leave `dev` as `0`, while `fstat()` on the open descriptor reports the real volume serial. The two settings guards required `descriptorStat.dev === pathStat.dev`, so every managed settings write failed closed and each failure left a stale `settings.json.ecc.lock` behind. This is the same defect class #2637 fixed in the memory vault guard, and the fix uses the same rule: compare `dev` only when both stats report one. The inode, regular-file, and `isSymbolicLink()` checks are unchanged, so the guards still fail closed on an actual swap. POSIX always reports `dev`, so behaviour on Linux/macOS is identical to before.

The fix is platform-agnostic rather than gated on `process.platform === 'win32'`, matching the memory vault helper, so a runtime that omits `dev` elsewhere would not reintroduce the lockout.

## Testing Done
- [x] Manual testing completed — reproduced the issue's failure sequence deterministically by stubbing `fs.lstatSync` to zero `dev` (see the new test), on macOS with Node 20.20.2. I do not have a Windows machine; the reporter confirmed on Windows 11 / Node 22.14.0 that `tests/lib/claude-settings.test.js` fails 6/36 on `5064474d` and passes 36/36 with this exact `sameDeviceId` rule applied.
- [x] Automated tests pass locally (`node tests/run-all.js`) — see the gauntlet result below.
- [x] Edge cases considered and tested — differing non-zero `dev` still rejected; BigInt stats in the lock module (`0n` is falsy, so the same rule covers both).

Targeted runs, all green: `tests/lib/claude-settings.test.js` (38/38), `tests/lib/claude-settings-array.test.js` (7/7), `tests/scripts/install-apply.test.js` (42/42), `tests/lib/install-lifecycle.test.js` (66/66), `tests/lib/install-executor.test.js` (20/20), `tests/lib/hook-consent.test.js` (8/8), `tests/lib/install-state.test.js` (6/6). `eslint` and `markdownlint` clean on touched files.

`node tests/run-all.js`: 4360 tests, 4353 passed, 7 failed. The 7 failures are in `tests/hooks/detect-project-worktree.test.js` (5) and `tests/hooks/observe-subdirectory-detection.test.js` (2) and fail identically on unmodified `main` in my environment: my login shell profile prints an SSH-agent banner into bash stdout, which those tests compare byte-for-byte. They do not touch the settings guards.

## Type of Change
- [x] `fix:` Bug fix

## Security & Quality Checklist
- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly (none touched)
- [x] Shell scripts pass shellcheck (not applicable)
- [x] Pre-commit hooks pass locally (if configured)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## If you changed dependencies or `package.json` (`bin` / `files` / deps)
Not applicable.

## If you added a skill, command, agent, hook, or CLI tool
Not applicable.

## Documentation
- [x] Updated relevant documentation (CHANGELOG)
- [x] Added comments for complex logic
- [ ] README updated (not needed)

---
Prepared with AI assistance (Claude Code); reviewed and tested locally by @apetcu.

### CODE REVIEW

File: tests/lib/claude-settings.test.js [hunk 1]
```diff
@@ -453,6 +453,59 @@ function runTests() {
...
+    // Node 22.12-22.16 and 24.0-24.1 on Windows report dev = 0 from path-based
+    // lstat() while fstat() on the open descriptor reports the volume serial.
+    const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'claude-settings-zero-dev-'));
+    const settingsPath = path.join(tempDir, 'settings.json');
+    const originalLstat = fs.lstatSync;
+    try {
+      fs.writeFileSync(settingsPath, '{"theme":"initial"}\n');
+      fs.lstatSync = function(target, options) {
+        const stats = originalLstat.call(fs, target, options);
+        stats.dev = options && options.bigint ? 0n : 0;
```
[the remark this answers is not in the document]
Comment (apetcu) [reply tests/lib/claude-settings.test.js]: Addressed in 55caaed8: the stubs now return a derived object via Object.create so the fs.Stats instance is left untouched and its methods keep working.

File: tests/lib/claude-settings.test.js [hunk 2]
```diff
@@ -453,6 +453,59 @@ function runTests() {
...
+    const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'claude-settings-other-dev-'));
+    const settingsPath = path.join(tempDir, 'settings.json');
+    const originalLstat = fs.lstatSync;
+    try {
+      fs.writeFileSync(settingsPath, '{"theme":"initial"}\n');
+      fs.lstatSync = function(target, options) {
+        const stats = originalLstat.call(fs, target, options);
+        if (path.resolve(String(target)) === settingsPath) {
+          stats.dev = options && options.bigint ? stats.dev + 1n : stats.dev + 1;
+        }
```
[the remark this answers is not in the document]
Comment (apetcu) [reply tests/lib/claude-settings.test.js]: Addressed in 55caaed8: stubs derive a new stats object instead of mutating the one lstatSync returned.

File: scripts/lib/install/claude-settings-lock.js [hunk 3]
```diff
@@ -7,7 +7,13 @@ const path = require('path');
...
 
 function sameFileIdentity(left, right) {
-  return left.dev === right.dev && left.ino === right.ino;
+  if (left.ino !== right.ino) return false;
+  // libuv 1.49.0 through 1.50.x (Node 22.12-22.16 and 24.0-24.1) resolve
+  // path-based lstat() on Windows without setting the volume serial, while
+  // fstat() on the open descriptor reports it. Compare dev only when both
+  // sides report one so the lock can still be released on those runtimes.
+  if (!left.dev || !right.dev) return true;
+  return left.dev === right.dev;
```
[the remark this answers is not in the document]
Comment (apetcu) [reply scripts/lib/install/claude-settings-lock.js:16]: Addressed in 55caaed8: added two tests that pin the strict branch — lock release refuses a quarantined lock whose lstat reports a different non-zero dev, and stale-lock recovery keeps a dead owner's lock under the same mismatch. Both fail if sameFileIdentity stops comparing dev.
